libminifi/src/utils/Cron.cpp

```diff
@@ -0,0 +1,510 @@
+    auto last_day = date.year()/date.month()/last;
+    auto target_date = local_days(last_day)-offset_;
+    return local_days(date) == target_date;
+  }
+
+ private:
+  days offset_;
+};
+
+class NthWeekdayField : public CronField {
+ public:
+  NthWeekdayField(weekday wday, uint8_t n) : weekday_(wday), n_(n) {}
+
+  [[nodiscard]] bool matches(local_seconds tp) const override {
+    year_month_day date(floor<days>(tp));
+    auto target_date = date.year()/date.month()/(weekday_[n_]);
+    return local_days(date) == local_days(target_date);
+  }
+
+ private:
+  weekday weekday_;
+  uint8_t n_;
+};
+
+class LastWeekDayField : public CronField {
+ public:
+  LastWeekDayField() = default;
+
+  [[nodiscard]] bool matches(local_seconds value) const override {
+    year_month_day date(floor<days>(value));
+    year_month_day last_day_of_the_month_date = year_month_day(local_days(date.year()/date.month()/last));
+    if (isWeekday(last_day_of_the_month_date))
+      return date == last_day_of_the_month_date;
+    year_month_day last_friday_of_the_month_date = year_month_day(local_days(date.year()/date.month()/Friday[last]));
+    return date == last_friday_of_the_month_date;
+  }
+};
+
+class LastSpecificDayOfTheWeekOfTheMonth : public CronField {
+ public:
+  explicit LastSpecificDayOfTheWeekOfTheMonth(weekday wday) : weekday_(wday) {}
+
+  [[nodiscard]] bool matches(local_seconds value) const override {
+    year_month_day date(floor<days>(value));
+    year_month_day last_weekday_of_month_date = year_month_day(local_days(date.year()/date.month()/weekday_[last]));
+    return date == last_weekday_of_month_date;
+  }
+ private:
+  weekday weekday_;
+};
+
+class ClosestWeekdayToTheNthDayOfTheMonth : public CronField {
+ public:
+  explicit ClosestWeekdayToTheNthDayOfTheMonth(day day_number) : day_number_(day_number) {}
+
+  [[nodiscard]] bool matches(local_seconds value) const override {
+    year_month_day date(floor<days>(value));
+    for (auto diff : {0, -1, 1, -2, 2}) {
+      auto target_date = date.year() / date.month() / (day_number_ + days(diff));
+      if (target_date.ok() && isWeekday(target_date))
+        return target_date == date;
+    }
+
+    return false;
+  }
+
+ private:
+  day day_number_;
+};
+
+template <typename FieldType>
+std::unique_ptr<CronField> parseCronField(const std::string& field_str) {
+  try {
+    if (field_str == "*") {
+      return std::make_unique<AllValuesField>();
+    }
+
+    if (field_str == "?") {
+      return std::make_unique<NotCheckedField>();
+    }
+
+    if (field_str == "L") {
+      if (std::is_same<day, FieldType>())
+        return std::make_unique<LastNthDayInMonthField>(days(0));
+      if (std::is_same<weekday, FieldType>())
+        return std::make_unique<SingleValueField<weekday>>(Saturday);
+      throw BadCronExpression("L can only be used in the Day of month/Day of week fields");
+    }
+
+    if (field_str == "LW") {
+      if (!std::is_same<day, FieldType>())
```

Comment:
Consider using `if constexpr` for these checks. It's not necessary, since all branches can be compiled in all cases, but makes it clearer that this is happening in compile-time.

Reply:
Good idea, I've added constexpr to the std::is_same ifs https://github.com/apache/nifi-minifi-cpp/pull/1335/commits/a7db03ebb2a795bff86aff210e4e29bcf424c372#